chemprop/data/v2/datasets.py

```diff
@@ -0,0 +1,263 @@
+        """Normalizes the features of the dataset using a :class:`~chemprop.data.StandardScaler`.
+
+        The :class:`~chemprop.data.StandardScaler` subtracts the mean and divides by the standard
+        deviation for each feature independently.
+
+        If a :class:`~chemprop.data.StandardScaler` is provided, it is used to perform the
+        normalization. Otherwise, a :class:`~chemprop.data.StandardScaler` is first fit to the
+        features in this  dataset and is then used to perform the normalization.
+
+        :param scaler: A fitted :class:`~chemprop.data.StandardScaler`. If it is provided it is
+            used, otherwise a new :class:`~chemprop.data.StandardScaler` is first fitted to this
+            data and is then used.
+        :param replace_nan_token: A token to use to replace NaN entries in the features.
+        :param scale_atom_descriptors: If the features that need to be scaled are atom features
+            rather than molecule.
+        :param scale_bond_features: If the features that need to be scaled are bond descriptors
+            rather than molecule.
+        :return: A fitted :class:`~chemprop.data.StandardScaler`. If a :class:`~chemprop.data.
+            StandardScaler` is provided as a parameter, this is the same :class:`~chemprop.data.
+            StandardScaler`. Otherwise, this is a new :class:`~chemprop.data.StandardScaler` that
+            has been fit on this dataset.
+        """
+        if len(self.data) == 0 or (
+            self.data[0].features is None and not scale_bond_features and not scale_atom_descriptors
+        ):
+            return None
+
+        if scaler is None:
+            if scale_atom_descriptors and not self.data[0].atom_descriptors is None:
+                features = np.vstack([d.raw_atom_descriptors for d in self.data])
+            elif scale_atom_descriptors and not self.data[0].atom_features is None:
+                features = np.vstack([d.raw_atom_features for d in self.data])
+            elif scale_bond_features:
+                features = np.vstack([d.raw_bond_features for d in self.data])
+            else:
+                features = np.vstack([d.raw_features for d in self.data])
+
+            scaler = StandardScaler(replace_nan_token=replace_nan_token)
+            scaler.fit(features)
+
+        if scale_atom_descriptors and not self.data[0].atom_descriptors is None:
+            for d in self.data:
+                d.atom_descriptors = scaler.transform(d.raw_atom_descriptors)
+        elif scale_atom_descriptors and not self.data[0].atom_features is None:
+            for d in self.data:
+                d.atom_features = scaler.transform(d.raw_atom_features)
+        elif scale_bond_features:
+            for d in self.data:
+                d.bond_features = scaler.transform(d.raw_bond_features)
+        else:
+            for d in self.data:
+                d.features = scaler.transform(d.raw_features.reshape(1, -1))[0]
+
+        return scaler
```

Comment:
I know the extra atom/bond features are currently not supported for reactions. This capability is expected to be supported in the future. Hence, I would suggest moving this part to MolGraphDataset.

Reply:
We can move it over when the features are actually supported. Until then, I don't want to add capability for something that's not actually usable